presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -21,88 +22,372 @@
+                    .orElseThrow(() -> new PrestoException(CONFIGURATION_INVALID, "Failed to load router config"));
+            Map<String, GroupSpec> newGroups = ImmutableMap.copyOf(newRouterSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
+            List<SelectorRuleSpec> newGroupSelectors = ImmutableList.copyOf(newRouterSpec.getSelectors());
+            Scheduler newScheduler = new SchedulerFactory(newRouterSpec.getSchedulerType()).create();
+            SchedulerType newSchedulerType = newRouterSpec.getSchedulerType();
+
+            List<URI> updatedAllClusters = newGroups.values().stream()
+                    .flatMap(groupSpec -> groupSpec.getMembers().stream())
+                    .collect(toImmutableList());
+
+            Map<URI, URI> newDiscoveryURIs = new HashMap<>();
+            this.initializeMembersDiscoveryURI(newDiscoveryURIs, newGroups);
+
+            updatedAllClusters.forEach(uri -> {
+                if (!remoteClusterInfos.containsKey(uri)) {
+                    log.info("Attaching cluster %s to the router", uri.getHost());
+                    remoteClusterInfos.put(uri, remoteInfoFactory.createRemoteClusterInfo(newDiscoveryURIs.get(uri)));
+                    remoteQueryInfos.put(uri, remoteInfoFactory.createRemoteQueryInfo(newDiscoveryURIs.get(uri)));
+                    log.info("Successfully attached cluster %s to the router. Queries will be routed to cluster after successful health check", uri.getHost());
+                }
+            });
+
+            for (URI uri : remoteClusterInfos.keySet()) {
+                if (!updatedAllClusters.contains(uri)) {
+                    log.info("Removing cluster %s from the router", uri.getHost());
+                    remoteClusterInfos.remove(uri);
+                    remoteQueryInfos.remove(uri);
+                    log.info("Successfully removed cluster %s from the router", uri.getHost());
+                }
+            }
+            this.currentConfig.set(new ClusterManagerConfig(newGroups, newGroupSelectors, newScheduler, newSchedulerType));
+        };
+        onConfigChangeDetection.apply();
+    }
 
-        this.groups = ImmutableMap.copyOf(routerSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
-        this.groupSelectors = ImmutableList.copyOf(routerSpec.getSelectors());
-        this.schedulerType = routerSpec.getSchedulerType();
-        this.scheduler = new SchedulerFactory(routerSpec.getSchedulerType()).create();
+    @PostConstruct
+    public void startConfigReloadTaskFileWatcher()
+    {
+        CompletableFuture.supplyAsync(() -> {
+            try (WatchService watchService = FileSystems.getDefault().newWatchService()) {
+                File routerConfigFile = new File(routerConfig.getConfigFile());
+                log.info(String.format("Router config watch service monitoring %s", routerConfig.getConfigFile()));
```

Comment:
don't need to do string format on the log statement
```suggestion
                log.info("Router config watch service monitoring %s", routerConfig.getConfigFile());
```